### ISSUE TITLE
directed_laplacian_matrix sometimes returns nans

### ISSUE THREAD

**Issue by jlazovskis** [networkx/networkx#6865]
The function `directed_laplacian_matrix` returns a matrix with `nan` entries on specific input graphs. The entries of the directed Laplacian matrix are always well-defined, so `nan`s should not appear. This behavior does not produce an error, but the code of the function produces a `RuntimeWarning`, and seems to come from the attempt to take the square root (with `numpy`) of an array with negative entries (produced by `scipy`).  

### Current Behavior

The function `directed_laplacian_matrix` in line `257` calls `scipy.sparse.linalg.eigs` to get the unique eigenvector whose eigenvalue is 1. This eigenvector should have all positive or all negative values,  but it seems that `scipy` sometimes returns some positive and some negative entries. The existence of a unique left eigenvector with all positive entries (corresponding to the eigenvalue 1) is guaranteed by the Perron--Frobenius theorem. The correct eigenvector is selected by `sparse.linalg.eigs` from `scipy`, and even if all the entries are negative (still a legitimate eigenvector), normalizing on line `259` should ensure all entries are positive. However, if the signs of entries are both positve and negative, then that will stay the case after this line.  When taking the square root later in line `260` with `numpy.sqrt`, a `RuntimeWarning` occurs, because the input is not necessarily positive. The warning is not produced when executing the function `directed_laplacian_matrix` directly, only when `np.sqrt(p)` is called on line `260` by itself. 

### Expected Behavior

It should be that `directed_laplacian_matrix` returns a matrix that does not have any `nan`s. Or, perhaps it should give an error if the vector `evecs` returned from `scipy` on line `257` does not have all positive or all negative entries. 

### Steps to Reproduce

I was trying to compute the directed Laplacian matrix for a directed graph with 408 nodes, accessible in `adjlist` format in [theattached file](https://github.com/networkx/networkx/files/12395264/G.txt) file (and also at a [pastebin link](https://pastebin.com/ET1k7vBq)) . This is the largest strongly connected component of a larger directed graph with 440 nodes (so the node indices go up to 440). The following code demonstrates that sometimes `nan` values are returned. The linked file is saved as `G.txt`.

````
import networkx as nx
G = nx.read_adjlist('G.txt', create_using=nx.DiGraph)
tries = 100
no_nan,has_nan = (0,0)
for _ in range(tries):
    mat = nx.directed_laplacian_matrix(G)
    nans = np.where(np.isnan(np.array(mat)))
    if len(nans[0] == 0):
        no_nan += 1
    else:
        has_nan += 1
print('In '+str(tries)+' attempts, '+str(no_nan)+' attempts with no nan values, '+str(has_nan)+' attempts with nan values', flush=True)

>>> In 100 attempts, 69 attempts with no nan values, 31 attempts with nan values
````

The following code of lines `251` to `260` shows that the `RuntimeWarning` happens on the `np.sqrt` line.

````
import numpy as np
import scipy as sp
import warnings
warnings.filterwarnings("error")
 
tries = 100
success,fail = (0,0)
for _ in range(tries):
    P = nx.linalg.laplacianmatrix._transition_matrix(G)
    evals,evecs = sp.sparse.linalg.eigs(P.T,k=1)
    v = evecs.flatten().real
    p = v/v.sum()
    try:
        sqrtp = np.sqrt(p)
        success += 1
    except:
        fail += 1
print('In '+str(tries)+' attempts, '+str(success)+' successes and '+str(fail)+' failures', flush=True)

>>> In 100 attempts, 27 successes and 73 failures
````

The precise warning is as follows:

````
RuntimeWarning: invalid value encountered in sqrt
  sqrtp = np.sqrt(p)
````

### Environment

Python version: 3.8.3
NetworkX version: 3.1
NumPy version: 1.23.1
SciPy version: 1.8.0

### Other comments

My guess is the issue is with rounding in `scipy`, so perhaps it is better to have this discussion in that project. However, it would be nice for `networkx` to provide some indication (error, warning) if this happens - now I only get an error if I try to do something with the returned matrix. There is no check made in `directed_laplacian_matrix` to see if the entries of `p` (the vector passed to the square root function) are positive, perhaps checking there and rasing error would better inform the user if there is a problem.

**Comment by dschult** [comment]
Thank you for this!
And for the (fairly small) example that shows the error message. I'd love an even smaller example to include in the tests, but this allowed me to trouble-shoot and understand the problem.

As you say, this is a problem with round-off switching the sign of an entry in the eigenvector and then taking the sqrt of the result. To fix it we should use `np.maximum(p, 0.)` or `np.abs(p)` before taking the sqrt. Quick tests confirm the internet claim that `np.abs(p)` is faster. Are there other reasons to choose `np.maximum`?

I put a fix in #6866

**Comment by jlazovskis** [comment]
I'm unsure how to make a smaller example, but I will think about it. The graph is quite large, so I'm having a hard time understanding its properties (and random node removal doesn't seem to work with less than ~380 nodes).

Regarding `np.abs` or `np.maximum`, it is more appropriate to use `np.abs`, because sometimes scipy returns the negative of the correct positive vector (in which case `np.maximum` would give a vector of (almost) all zeros). 

Additionally, I suggest putting `np.abs` before computing `p = v/v.sum()`,  as the difference in sign would make this not a true normalization (if `v` has both positive and negative entries).

**Comment by dschult** [comment]
I found examples as small as 50 nodes by taking your example and repeatedly removing one node at a time while checking to see if the nan's still showed up.  The problem is that round-off is not reliable/repeatable/predictable so even with the small examples, it often doesn't happen that the round-off switches the sign of the values close to zero.

In that process I came to understand that this isn't really a test-able issue in the sense of CI / unit tests.  It is sufficient to make sure it doesn't happen without testing that it doesn't happen. So the PR is merged now.  As far as placement of the `np.abs`, the difference will only matter at the level of round-off which is inherent in every step anyway. So normalization is only "true" up to round-off error anyway. 

Thanks --
